src/server/raft_client.rs

```diff
@@ -12,35 +12,39 @@
 // limitations under the License.
 
 use std::ffi::CString;
-use std::sync::atomic::{AtomicBool, AtomicI32, Ordering};
-use std::sync::Arc;
-
-use futures::sync::mpsc::{self, UnboundedSender};
-use futures::sync::oneshot::{self, Sender};
-use futures::{stream, Future, Sink, Stream};
-use grpc::{ChannelBuilder, Environment, WriteFlags};
+use std::i64;
+use std::sync::atomic::{AtomicI32, Ordering};
+use std::sync::{Arc, Mutex};
+use std::time::Instant;
+
+use crossbeam::channel::SendError;
+use futures::{future, stream, Future, Poll, Sink, Stream};
+use grpc::{ChannelBuilder, Environment, Error as GrpcError, RpcStatus, RpcStatusCode, WriteFlags};
 use kvproto::raft_serverpb::RaftMessage;
+use kvproto::tikvpb::BatchRaftMessage;
 use kvproto::tikvpb_grpc::TikvClient;
+use protobuf::RepeatedField;
+use tokio::runtime::Runtime;
+use tokio::timer::Delay;
 
+use super::load_statistics::ThreadLoad;
 use super::metrics::*;
-use super::{Config, Error, Result};
-use util::collections::HashMap;
+use super::{Config, Result};
+use util::collections::{HashMap, HashMapEntry};
+use util::mpsc::batch::{self, Sender as BatchSender};
 use util::security::SecurityManager;
 
 const MAX_GRPC_RECV_MSG_LEN: i32 = 10 * 1024 * 1024;
 const MAX_GRPC_SEND_MSG_LEN: i32 = 10 * 1024 * 1024;
-const PRESERVED_MSG_BUFFER_COUNT: usize = 1024;
+
+const RAFT_MSG_MAX_BATCH_SIZE: usize = 128;
+const RAFT_MSG_NOTIFY_SIZE: usize = 8;
```

Comment:
Have you tried different values for the above two variables?

Reply:
I have tried 8/64, 16/64 and 16/128, not found obvious difference between them.